scripts/install_backend_python_libs.py

```diff
@@ -0,0 +1,364 @@
+           listed in the requirements file.
+
+    Returns:
+        dict(string, tuple(string|None, string|None)). Dictionary with the
+        library names as keys and tuples as values. The 1st element of the
+        tuple is the version string of the library required by the
+        requirements.txt file while the 2nd element is the version string of
+        the library currently in the 'third_party/python_libs' directory. If
+        the library doesn't exist, the corresponding tuple element will be None.
+        For example, the following dictionary signifies that 'requirements.txt'
+        requires flask with version 1.0.1 while the 'third_party/python_libs'
+        directory contains flask 1.1.1 (or mismatch 4 above):
+            {
+              flask: ('1.0.1', '1.1.1')
+            }
+    """
+    requirements_contents = _get_requirements_file_contents()
+    directory_contents = _get_third_party_python_libs_directory_contents()
+
+    mismatches = {}
+    for library_name in requirements_contents:
+        # Library exists in the directory and the requirements file.
+        if library_name in directory_contents:
+            # Library matches but version doesn't match.
+            if (directory_contents[library_name] !=
+                    requirements_contents[library_name]):
+                mismatches[library_name] = (
+                    requirements_contents[library_name],
+                    directory_contents[library_name])
+        # Library exists in the requirements file but not in the directory.
+        else:
+            mismatches[library_name] = (
+                requirements_contents[library_name], None)
+
+    for library_name in directory_contents:
+        # Library exists in the directory but is not in the requirements file.
+        if library_name not in requirements_contents:
+            mismatches[library_name] = (None, directory_contents[library_name])
+
+    return mismatches
+
+
+def validate_metadata_directories():
+    """Validates that for each installed library in the
+    'third_party/python_libs' folder, there exists corresponding metadata
+    directories following the correct naming conventions that are detailed by
+    the PEP-427 and PEP-376 python guidelines.
+
+    Raises:
+        Exception. An installed library's metadata does not exist the
+            'third_party/python_libs' directory in the format that we expect
+            (following the PEP-427 and PEP-376 python guidelines).
+    """
+    directory_contents = _get_third_party_python_libs_directory_contents()
+    directory_names = set(
+        [
+            name.lower()
+            for name in os.listdir(common.THIRD_PARTY_PYTHON_LIBS_DIR)
+        ])
+    for library_name, version_string in directory_contents.items():
+        # Possible names of the metadata directory installed.
+        possible_file_names = _get_possible_metadata_directory_names(
```

Comment:
I'm having a lot of trouble understanding what's going on in this function. Perhaps you could simplify it.

- Why are we doing .lower() above? Needs explanation (see a previous comment).
- You're using file names to talk about dir names.
- I don't understand what "has name in directory" means.
- Maybe, below, you could do a one-liner: `if any([name in directory_names for name in possible_file_names])`

Reply:
Added more explanation and also explained most of it a previous comment about lower. I changed it to be more clear and changed the if statement to a one-liner.